File: frontend/lib/usuario.ts
// Cliente de la cuenta liviana del comprador/visitante: solo celular + código
// por WhatsApp, sin contraseña. Objetivo único: guardar comercios favoritos
// y dejar el celular con consentimiento para avisos/ofertas. No confundir
// con las cuentas de comercio (lib/comercio.ts).
const API = process.env.NEXT_PUBLIC_API_URL ?? "http://localhost:8000";
const TOKEN_KEY = "bermejo_usuario_token";
const SESSION_KEY = "bermejo_usuario";

export type UsuarioSession = { id: string; whatsapp: string };

export function getUsuarioToken(): string | null {
  return typeof window === "undefined" ? null : localStorage.getItem(TOKEN_KEY);
}
export function getUsuarioSession(): UsuarioSession | null {
  if (typeof window === "undefined") return null;
  const raw = localStorage.getItem(SESSION_KEY);
  return raw ? (JSON.parse(raw) as UsuarioSession) : null;
}
export function clearUsuario() {
  localStorage.removeItem(TOKEN_KEY);
  localStorage.removeItem(SESSION_KEY);
}

export async function solicitarCodigoUsuario(whatsapp: string): Promise<void> {
  const res = await fetch(`${API}/auth/usuario/solicitar-codigo`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ whatsapp }),
  });
  if (!res.ok) throw new Error("No se pudo enviar el código");
}

export async function verificarCodigoUsuario(whatsapp: string, codigo: string): Promise<UsuarioSession> {
  const res = await fetch(`${API}/auth/usuario/verificar`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ whatsapp, codigo }),
  });
  if (!res.ok) {
    const d = await res.json().catch(() => ({}));
    throw new Error(d.detail ?? "Código incorrecto");
  }
  const data = await res.json();
  localStorage.setItem(TOKEN_KEY, data.access_token);
  localStorage.setItem(SESSION_KEY, JSON.stringify(data.usuario));
  return data.usuario as UsuarioSession;
}

async function uFetch(path: string, init?: RequestInit) {
  const res = await fetch(`${API}${path}`, {
    ...init,
    headers: {
      "Content-Type": "application/json",
      Authorization: `Bearer ${getUsuarioToken() ?? ""}`,
      ...(init?.headers ?? {}),
    },
  });
  if (res.status === 401) {
    clearUsuario();
    throw new Error("Tu sesión venció, volvé a entrar");
  }
  if (!res.ok) throw new Error("Error en la solicitud");
  return res.json();
}

export type FavoritoComercio = {
  id: string; slug: string; nombre: string; logo_url: string | null; portada_url: string | null;
  direccion: string | null; rating: number; whatsapp: string; verificado: boolean;
};

export const listarFavoritos = (): Promise<FavoritoComercio[]> =>
  uFetch("/usuario/favoritos").then((d) => d.items as FavoritoComercio[]);

export const agregarFavorito = (comercioId: string): Promise<void> =>
  uFetch("/usuario/favoritos", { method: "POST", body: JSON.stringify({ comercio_id: comercioId }) });

export const quitarFavorito = (comercioId: string): Promise<void> =>
  uFetch(`/usuario/favoritos/${comercioId}`, { method: "DELETE" });

File: backend/app/api/usuario.py
"""Cuenta liviana del comprador/visitante: solo celular + código por WhatsApp,
sin contraseña. Objetivo único: capturar el número con consentimiento para
avisos/ofertas, y permitir guardar comercios favoritos desde cualquier
dispositivo. No tiene nada que ver con las cuentas de comercio."""
import secrets
from datetime import datetime, timedelta, timezone

import structlog
from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel

from app.core import auth
from app.core.config import settings
from app.db.repository import Repo, get_repo

router = APIRouter()
logger = structlog.get_logger()


class SolicitarCodigoBody(BaseModel):
    whatsapp: str


@router.post("/auth/usuario/solicitar-codigo")
def solicitar_codigo(body: SolicitarCodigoBody, repo: Repo = Depends(get_repo)) -> dict:
    """Crea el usuario si no existe (alta progresiva) y genera un código para
    que lo confirme mandando un WhatsApp entrante (no lo mandamos nosotros —
    ver docs/pendientes.md sección 0, evita riesgo de ban por envío saliente)."""
    if not body.whatsapp.strip():
        raise HTTPException(status_code=400, detail="Falta el número de WhatsApp")
    usuario = repo.get_usuario_por_whatsapp(body.whatsapp) or repo.crear_usuario(body.whatsapp)
    code = f"{secrets.randbelow(1_000_000):06d}"
    expira = (datetime.now(timezone.utc) + timedelta(minutes=15)).isoformat()
    repo.set_reset_code_usuario(usuario["id"], code, expira)
    logger.info("usuario.solicitar_codigo", whatsapp=body.whatsapp)
    return {"ok": True, "codigo": code, "wa_link": settings.wa_link_confirmar(code)}


class VerificarBody(BaseModel):
    whatsapp: str
    codigo: str


@router.post("/auth/usuario/verificar")
def verificar(body: VerificarBody, repo: Repo = Depends(get_repo)) -> dict:
    """El frontend pollea esto después de mostrar el botón "Confirmar por
    WhatsApp" — falla con 400 hasta que el webhook marque el código como
    confirmado (mensaje entrante real), no alcanza con que el código matchee."""
    usuario = repo.get_usuario_por_whatsapp(body.whatsapp)
    if not usuario or not usuario.get("reset_code") or usuario["reset_code"] != body.codigo:
        raise HTTPException(status_code=400, detail="Código incorrecto")
    expira = usuario.get("reset_code_expira")
    if not expira or datetime.fromisoformat(expira) < datetime.now(timezone.utc):
        raise HTTPException(status_code=400, detail="El código venció, pedí uno nuevo")
    if not usuario.get("reset_code_confirmado_at"):
        raise HTTPException(status_code=400, detail="Todavía no confirmaste por WhatsApp")
    repo.set_reset_code_usuario(usuario["id"], None, None)
    token = auth.make_usuario_token(usuario["id"], usuario["whatsapp"])
    logger.info("usuario.verificado", whatsapp=body.whatsapp)
    return {"access_token": token, "usuario": {"id": usuario["id"], "whatsapp": usuario["whatsapp"]}}


@router.get("/usuario/favoritos")
def listar_favoritos(claims: dict = Depends(auth.require_usuario), repo: Repo = Depends(get_repo)) -> dict:
    items = repo.list_favoritos(claims["usuario_id"])
    return {"items": [i["comercios"] for i in items if i.get("comercios")]}


class FavoritoBody(BaseModel):
    comercio_id: str


@router.post("/usuario/favoritos")
def agregar_favorito(body: FavoritoBody, claims: dict = Depends(auth.require_usuario), repo: Repo = Depends(get_repo)) -> dict:
    repo.agregar_favorito(claims["usuario_id"], body.comercio_id)
    return {"ok": True}


@router.delete("/usuario/favoritos/{comercio_id}")
def quitar_favorito(comercio_id: str, claims: dict = Depends(auth.require_usuario), repo: Repo = Depends(get_repo)) -> dict:
    repo.quitar_favorito(claims["usuario_id"], comercio_id)
    return {"ok": True}
